common/src/jni/main/cpp/conscrypt/jniutil.cc

```diff
@@ -124,6 +124,7 @@ bool isGetByteArrayElementsLikelyToReturnACopy(size_t size) {
 
 int jniThrowException(JNIEnv* env, const char* className, const char* msg) {
     jclass exceptionClass = env->FindClass(className);
+    ERR_clear_error();
```

Comment:
I think this is rather awkward, and may still be buggy for errors that we decide not to throw exceptions for. Would it be better to have a ScopedErrorClearer that calls ERR_clear_error in its destructor for methods that should never leave an error pending ?

Reply:
+1

Reply:
I think we want to be more explicit about when we clear the errors rather than just doing it wholesale every time we leave a method that could have generated one, since that makes it easy to miss in the other direction and just ignore an error without seeing it.  Especially because "a method that could have generated one" is pretty much every native method we have.

How do you feel about doing explicit clearing but adding a class that asserts that the error stack is clear in its destructor?  That way we could find any case where we're failing to properly handle an error.  We have the option of compiling it out in release builds as well to avoid any performance impact.

Reply:
Sure, but even so, having errors cleared in jniThrowException seems kludgy to me. Throwing a Java exception seems unrelated to clearing an error - exceptions might be thrown for reasons other than BoringSSL errors. 

If you'd rather clear things explicitly, I think I prefer the older approach of explicitly calling clear in conjunction with your proposed check in the destructor.

Reply:
That's reasonable, yeah, I can put that together.